### PR TITLE
Add "For First" and "For Third" columns to forecast view

### PR DESCRIPTION
**Pull request by HaysCubing**
Add 2 new statistics columns, "For First" and "For Third" to forecast view.

Columns show the time needed for a competitor to overtake 1st/3rd on their next solve.

All calculations are done using projected averages.

### CODE REVIEW

**Comment by HaysCubing** [comment]
@jonatanklosko let me know what you think of this change. It's more drastic than the first "forecast view" change since it adds two new statics columns. I think this will be really nice to have for the projector display at worlds so people in the crowd don't have to do the math on what so and so needs to win / podium.

It should be pretty clean already, but wanted to get general ideas and comments out of the way now before I write tests and add documentation

File: client/src/lib/result.js [hunk 1]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
...
+          );
+        }
+      }
+    }
+  }
+
   return resultsForView;
 }
 
+export function timeNeededToOvertake(
```
Comment (HaysCubing) [inline client/src/lib/result.js]: Wasn't sure if parts of this should be functioned out to attempt-result.js

File: client/src/lib/result.js [hunk 2]
```diff
@@ -1,32 +1,48 @@
...
+  DNF_VALUE,
 } from "./attempt-result";
 
 /**
  * Returns a list of objects corresponding to result statistics - best and average.
  * The first statistic is the one that determines the ranking.
  * This is a common logic used in all result tables/dialogs.
  */
-export function orderedResultStats(eventId, format) {
+export function orderedResultStats(eventId, format, forecastView = null) {
```
Comment (jonatanklosko) [inline client/src/lib/result.js]:
```suggestion
export function orderedResultStats(eventId, format, forecastView = false) {
```

File: client/src/lib/result.js [hunk 3]
```diff
@@ -1,32 +1,48 @@
...
   const { numberOfAttempts, sortBy } = format;
 
   if (!shouldComputeAverage(eventId, numberOfAttempts)) {
     return [{ name: "Best", field: "best", recordTagField: "singleRecordTag" }];
   }
 
-  const stats = [
+  var stats = (stats = [
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: When defining variables, prefer to use `const` and `let` (when you reassign the name later). `var` is an old specifier and has its issues, which was the reason for introducing `const` and `let` :)

```suggestion
  let stats = [
```

(same for all other `var` usages)

File: client/src/lib/attempt-result.js [hunk 4]
```diff
@@ -124,8 +125,11 @@ function meanOfX(attemptResults) {
 }
 
 function mean(values) {
-  const sum = values.reduce((x, y) => x + y, 0);
-  return Math.round(sum / values.length);
+  return Math.round(sum(values) / values.length);
+}
+
+export function sum(values) {
+  return values.reduce((x, y) => x + y, 0);
```
Comment (jonatanklosko) [inline client/src/lib/attempt-result.js]: It's a detail, but I wouldn't export this function from this module, because it's not specific to attempt results at all. We could extract to utils, but I don't think it's worth it either, I would just redefine the function in the other module, just to avoid writing the `reduce` inline.

File: client/src/lib/result.js [hunk 5]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
     prevResult = currentResult;
   }
 
+  const secondComplete =
+    resultsForView.length > 1 && isComplete(resultsForView[1].projectedAverage);
+  const fourthComplete =
+    resultsForView.length > 3 && isComplete(resultsForView[3].projectedAverage);
+  if (secondComplete) {
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: I am missing why we check for second and fourth completeness. Could you explain?

Also note that 4th can be complete, even if 2nd is note, because we rerank based on projected average.

File: client/src/lib/result.js [hunk 6]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
...
+    for (let i = 0; i < resultsForView.length; i++) {
+      var result = resultsForView[i];
+      if (isSkipped(result.projectedAverage)) {
+        break;
+      }
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: I would make the same check we do in the prior loop, which is equivalent.

```suggestion
      if (currentResult.attempts.length === 0) {
        // From this point forward all results are empty, so we are done
        break;
      }
```

File: client/src/lib/result.js [hunk 7]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
...
+      if (isSkipped(result.average)) {
+        var firstIndex = i == 0 ? 1 : 0;
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: If this result is the first, I would probably just set `forFirst = 0`, because it just doesn't apply.

If we wanted to show DNF, I would perhaps do this:

```elixir
result.forFirst = timeNeededToOvertake(
  result,
  format,
  resultsForView[0].projectedAverage,
  resultsForView[0].best
);

if (i === 0 && !isSkipped(result.forFirst)) {
  result.forFirst = DNF_VALUE;
}
```

So if it's the first result, we use `timeNeededToOvertake` effectively to check if the stat is computed at all, in which case we set DNF, because we know we are already first.

But again, I think setting 0 should be totally fine.

(same for `forThird`)

File: client/src/lib/result.js [hunk 8]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
...
+  }
+
   return resultsForView;
 }
 
+export function timeNeededToOvertake(
+  result,
+  format,
+  overtakeAverage,
+  overtakeBest
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: Since you accept `result` as the argument, let's also accept `overtakeResult` for consistency (instead of passing average and best separately).

File: client/src/lib/result.js [hunk 9]
```diff
@@ -135,9 +153,93 @@ export function resultsForView(results, format, forecastView) {
...
+
+  const isMean = format.numberOfAttempts === 3 || result.attempts.length < 2;
+  const nextCountingSolves = result.attempts.length + (isMean ? 1 : -1);
+  const totalNeeded = overtakeAverage * nextCountingSolves;
+  // For a mean of 3, .01 can be added to achieve the same rounded result
+  const roundingBuffer = nextCountingSolves === 3 ? 1 : 0;
+  var countingSum = sum(attemptResults);
+  if (!isMean) {
+    countingSum =
+      countingSum - Math.min(...attemptResults) - Math.max(...attemptResults);
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: Min/max here needs to account for DNF (-1).

File: client/src/lib/result.js (hunk 1 above)
Comment (jonatanklosko) [reply client/src/lib/result.js]: Given that we deal with both best and average, I think it belongs here just fine.

Btw. you don't necessarily need to export this function, you could write tests for `resultsForView` and this function is just an implementation detail. But if testing this function makes it easier to cover edge cases, then that's fine as well.

**Review by jonatanklosko — COMMENTED** [review]
Hey @HaysCubing, thanks for working on this, here's a few comments.

We definitely need tests :D

File: client/src/lib/result.js (hunk 3 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: Ah didn't know that about var, let, and const. Thanks!

File: client/src/lib/result.js (hunk 1 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: You have to export functions to use them in tests in js? Interesting.

I didn't export for now - I'll reevaluate when I write tests

File: client/src/lib/result.js (hunk 5 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: Since we rerank on projected average, doesn't that mean if 4th is complete 2nd has to be complete as well? All of the completed projected averages will be listed first.

File: client/src/lib/result.js (hunk 7 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: So the idea here was that if the first/third positions aren't complete, we still want to see what time they need to stay in their current position (since their average could get worse on the next solve). I think this is pretty important to have since it covers the case where the projected winner is on their last solve and we still want to know what they need to secure the win.

I'm realizing it's going to make the logic a bit annoying here because 1st/3rd get compared with 2nd/4th.

File: client/src/lib/result.js (hunk 8 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: Yeah wow that's way cleaner

File: client/src/lib/result.js (hunk 5 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: Cleaned this up a lot. Now function `timeNeededToOvertake` can handle incomplete values and so we don't need to check if certain values are complete or not. Loop just needs to make sure it won't hit an out of bounds error.

**Comment by HaysCubing** [comment]
Still working on this iteration. `timeNeededToOvertake` is getting complicated since now it handles incomplete results and will definitely need to have rigorous tests.

Looks like you're on board with the general ideas though so I'll go ahead and write tests and document everything for my next iteration ready for review. Will have time later in the next week.

File: client/src/lib/result.js (hunk 7 above)
Comment (HaysCubing) [reply client/src/lib/result.js]: Actually not as complicated after I updated `timeNeededToOvertake` to handle skipped and incomplete values

File: client/src/lib/result.js (hunk 5 above)
Comment (jonatanklosko) [reply client/src/lib/result.js]: Ah, the check was on projectedAverage, not average, I get it.

File: client/src/lib/attempt-result.js [hunk 10]
```diff
@@ -303,6 +305,8 @@ export function formatAttemptResult(attemptResult, eventId) {
   if (attemptResult === SKIPPED_VALUE) return "";
   if (attemptResult === DNF_VALUE) return "DNF";
   if (attemptResult === DNS_VALUE) return "DNS";
+  if (attemptResult === NA_VALUE) return "N/A";
+  if (attemptResult === SUCCESS_VALUE) return "SUCCESS";
```
Comment (jonatanklosko) [inline client/src/lib/attempt-result.js]: I would prefer to only use the standardized values in this module (i.e. we don't expect -3 or -4 to ever be stored as an actual value in the database). I was on the fence about `NA_VALUE`, but now that there is `SUCCESS_VALUE` I think we should definitely do this elsewhere.

I think what I would do instead, is to have the `NA_VALUE` and `SUCCESS_VALUE` constants in `result.js`, and in that same file define `formatAttemptResultForN`. It would check for `NA_VALUE`/`SUCCESS_VALUE` first, and fallback to `formatAttemptResult` otherwise. In `ResultStat` component we can special case the `forFirst`/`forThird` fields to use `formatAttemptResultForN`.

Does that make sense?

File: client/vite.config.js [hunk 11]
```diff
@@ -17,6 +17,6 @@ export default defineConfig({
     },
     // Uncomment when running using Docker on Windows.
     // See https://github.com/vitejs/vite/issues/1153#issuecomment-785467271
-    // watch: { usePolling: true },
+    watch: { usePolling: true },
```
Comment (HaysCubing) [inline client/vite.config.js]: Undo this change before submitting

File: client/src/lib/attempt-result.js (hunk 10 above)
Comment (HaysCubing) [reply client/src/lib/attempt-result.js]: Yep sounds good!

I added function `formatAttemptResultForN` and just changed all references in `ResultStat.jsx` to use it. Since it's just a wrapper around `formatAttemptResult` I figured it would be cleaner and functionally the same but let me know if I should add special logic in `ResultStat.jsx` to call the different versions.

File: client/src/lib/tests/result.test.js [hunk 12]
```diff
@@ -148,4 +188,341 @@ describe("viewResults", () => {
...
+});
+
+describe("timeNeededToOvertake", () => {
+  it("returns DNF if overtake result is skipped", () => {
+    expect(timeNeededToOvertake(null, null, { projectedAverage: 0 })).toEqual(
+      -1
+    );
+  });
+
+  it("handles results turning from a mean to an average", () => {
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js]:
```suggestion
  it("handles projected average turning from a mean to an average", () => {
```

File: client/src/lib/tests/result.test.js [hunk 13]
```diff
@@ -148,4 +188,341 @@ describe("viewResults", () => {
...
+          best: 110,
+          projectedAverage: 110,
+          attempts: [{ result: 110 }, { result: 110 }],
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(81);
+    overtakeResult = { best: 50, projectedAverage: 100 };
+    expect(
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:425]: It would be helpful to have a short comment before each case outlining the case, for example "Can overtake with better mean and worse single", "Can overtake with same mean, and getting a better single", etc. I would also add an extra newline before the cases.

Those could be separate tests altogether, but it's also fine as is, it's just about making it easier for the reader to understand which scenario are we asserting (instead of reverse-figuring it out by analyzing the times).

File: client/src/lib/result.js [hunk 14]
```diff
@@ -63,13 +82,29 @@ export function forecastViewSupported(round) {
   );
 }
 
+/**
+ * Wrapper for formatAttemptResult. Also handles values
+ * NA_VALUE and SUCCESS_VALUE
+ */
+export function formatAttemptResultForN(attemptResult, eventId) {
```
Comment (jonatanklosko) [inline client/src/lib/result.js]: Since we are using it for in all cases, I would call it `formatAttemptResultForView`, which makes it clear it's a generic version.

**Review by jonatanklosko — COMMENTED** [review]
The logic looks correct to me, just a few minor comments and I think it's good to go :)

File: client/src/lib/tests/result.test.js (hunk 13 above)
Comment (HaysCubing) [reply client/src/lib/tests/result.test.js:425]: Yeah I was thinking the same - wasn't sure what the proper style was for these tests.

Added comments

Comment (HaysCubing) [reply client/src/lib/tests/result.test.js:425]: Adding newline seems unnecessary when there is a comment already - let me know if you want me to add that still

File: client/src/lib/tests/result.test.js [hunk 15]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+describe("timeNeededToOvertake", () => {
+  it("returns DNF if overtake result is skipped", () => {
+    expect(timeNeededToOvertake(null, null, { projectedAverage: 0 })).toEqual(
+      -1
+    );
+  });
+
+  it("handles projected average turning from a mean to an average", () => {
+    const format = { numberOfAttempts: 5 };
+    let result = { attempts: [{ result: 99 }, { result: 100 }], best: 99 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:259]:
```suggestion
    const format = { numberOfAttempts: 5 };
    let result = { attempts: [{ result: 99 }, { result: 100 }], best: 99 };

```

File: client/src/lib/tests/result.test.js [hunk 16]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+    // Worst possible average beats overtake average
+    expect(
+      timeNeededToOvertake(result, format, { best: 10, projectedAverage: 101 })
+    ).toEqual(-1);
+    // Worst possible average ties overtake average, wins on best
+    expect(
+      timeNeededToOvertake(result, format, { best: 100, projectedAverage: 100 })
+    ).toEqual(-1);
+    result = { attempts: [{ result: 100 }, { result: -1 }], best: 100 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:270]:
```suggestion

    result = { attempts: [{ result: 100 }, { result: -1 }], best: 100 };

```

File: client/src/lib/tests/result.test.js [hunk 17]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+    ).toEqual(-4);
+    // Best possible average ties overtake average
+    expect(
+      timeNeededToOvertake(result, format, { best: 50, projectedAverage: 100 })
+    ).toEqual(49);
+    // Best possible average loses to overtake average
+    expect(
+      timeNeededToOvertake(result, format, { best: 50, projectedAverage: 50 })
+    ).toEqual(-3);
+    result = { attempts: [{ result: -1 }, { result: -1 }], best: -1 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:293]:
```suggestion

    result = { attempts: [{ result: -1 }, { result: -1 }], best: -1 };

```

File: client/src/lib/tests/result.test.js [hunk 18]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+    // Best possible average ties overtake average, overtake average incomplete
+    expect(
+      timeNeededToOvertake(result, format, { best: -1, projectedAverage: -1 })
+    ).toEqual(-4);
+  });
+
+  it("handles incomplete overtake results", () => {
+    let overtakeResult = { best: 50, projectedAverage: -1 };
+    const format = { numberOfAttempts: 5 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:303]:
```suggestion
    const format = { numberOfAttempts: 5 };

```

File: client/src/lib/tests/result.test.js [hunk 19]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+        {
+          best: 100,
+          projectedAverage: 100,
+          attempts: [{ result: 100 }],
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(-4);
+    overtakeResult = { best: -1, projectedAverage: -1 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:360]:
```suggestion

    overtakeResult = { best: -1, projectedAverage: -1 };

```

File: client/src/lib/tests/result.test.js [hunk 20]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(-4);
+  });
+
+  it("handles incomplete results", () => {
+    const overtakeResult = { best: 50, projectedAverage: 50 };
+    const format = { numberOfAttempts: 3 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:378]:
```suggestion
    const format = { numberOfAttempts: 3 };

```

File: client/src/lib/tests/result.test.js [hunk 21]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(-3);
+  });
+
+  it("handles overtake for mean", () => {
+    let overtakeResult = { best: 100, projectedAverage: 100 };
+    let format = { numberOfAttempts: 3 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:395]:
```suggestion
    let format = { numberOfAttempts: 3 };

```

File: client/src/lib/tests/result.test.js [hunk 22]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+        {
+          best: 110,
+          projectedAverage: 110,
+          attempts: [{ result: 110 }, { result: 110 }],
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(81);
+    overtakeResult = { best: 50, projectedAverage: 100 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:422]:
```suggestion

    overtakeResult = { best: 50, projectedAverage: 100 };

```

File: client/src/lib/tests/result.test.js [hunk 23]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+          best: 110,
+          projectedAverage: 110,
+          attempts: [{ result: 110 }, { result: 110 }],
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(78);
+
+    overtakeResult = { best: 80, projectedAverage: 100 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:449]:
```suggestion
    overtakeResult = { best: 80, projectedAverage: 100 };

```

File: client/src/lib/tests/result.test.js [hunk 24]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+          projectedAverage: 110,
+          attempts: [{ result: 110 }, { result: 110 }],
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(79);
+
+    format = { numberOfAttempts: 5 };
+    overtakeResult = { best: 100, projectedAverage: 100 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:465]:
```suggestion
    overtakeResult = { best: 100, projectedAverage: 100 };

```

File: client/src/lib/tests/result.test.js [hunk 25]
```diff
@@ -148,4 +188,366 @@ describe("viewResults", () => {
...
+        },
+        format,
+        overtakeResult
+      )
+    ).toEqual(90);
+  });
+
+  it("handles overtake for average", () => {
+    let overtakeResult = { best: 50, projectedAverage: 100 };
+    const format = { numberOfAttempts: 5 };
```
Comment (jonatanklosko) [inline client/src/lib/tests/result.test.js:483]:
```suggestion
    const format = { numberOfAttempts: 5 };

```

**Review by jonatanklosko — COMMENTED** [review]
I would add newlines around the reassignments. I started reading and got confused about an `expect`, and only then realised the result is different. It is easy to miss with no spacing.

**Review by jonatanklosko — APPROVED** [review]
Thank you :)
